### PR TITLE
Fix analyzer - insertion from select with subquery referencing insertion table should process only insertion block.

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix analyzer - insertion from select with subquery referencing insertion table should process only insertion block.

Tests:

- [ ] 01155_rename_move_materialized_view
- [ ] 01268_mv_scalars
- [x] 01947_mv_subquery
- [x] 02139_MV_with_scalar_subquery
- [ ] 02174_cte_scalar_cache_mv
- [ ] 02479_race_condition_between_insert_and_droppin_mv

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit b680b1ac1ed8623e5dc57707a5fec74726f0c9fb with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50857/b680b1ac1ed8623e5dc57707a5fec74726f0c9fb/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Review by novikd — COMMENTED** [review]
In general LGTM, but this PR is not a bug-fix because it's a not-enabled experimental feature. You should update the PR description.

**Comment by novikd** [comment]
And if it fixes `01268_mv_scalars ` you can remove it from the `broken_tests.txt`

**Comment by novikd** [comment]
@yakov-olkhovskiy I thought a bit about this problem, and I think we can do a simple and elegant solution for it. What we actually want is to resolve the table identifier to the table that represents the insertion block. So actually we just need to modify `QueryAnalyzer::tryResolveTableIdentifierFromDatabaseCatalog` and build there a `TableNode` referencing the required table.

**Comment by yakov-olkhovskiy** [comment]
@novikd I don't see much difference with what I've done - anyway you will need to drag this values storage all the way through QueryAnalysisPass to QueryAnalyzer - my concern is that ugly solution of dragging this storage along the way...

**Comment by yakov-olkhovskiy** [comment]
...I just realized that we already have this storage in the context:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Transforms/buildPushingToViewsChain.cpp#L476-L487

File: src/Interpreters/InterpreterSelectQueryAnalyzer.cpp [hunk 1]
```diff
@@ -109,6 +109,9 @@ void replaceStorageInQueryTree(QueryTreeNodePtr & query_tree, const ContextPtr &
         }
     }
 
+    if (auto * table_node = table_expression_to_replace->as<TableNode>(); table_node && table_node->getStorageID().getFullNameNotQuoted() != storage->getStorageID().getFullTableName())
```
Comment (novikd) [inline src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:112]: It definitely requires a comment.

**Comment by KochetovNicolai** [comment]
02404_memory_bound_merging is flaky with analyzer

**Review by KochetovNicolai — APPROVED** [review]

**Pull request merged**